pulsar-client-cpp/lib/ConnectionPool.cc

```diff
@@ -24,34 +23,45 @@ namespace pulsar {
 
 ConnectionPool::ConnectionPool(const ClientConfiguration& conf,
                                ExecutorServiceProviderPtr executorProvider,
-                               const AuthenticationPtr& authentication, bool poolConnections)
+                               const AuthenticationPtr& authentication,
+                               bool poolConnections, size_t connectionsPerBroker)
         : clientConfiguration_(conf),
           executorProvider_(executorProvider),
           authentication_(authentication),
           pool_(),
           poolConnections_(poolConnections),
-          mutex_() {
+          mutex_(),
+          connectionsPerBroker(connectionsPerBroker) {
 }
 
 Future<Result, ClientConnectionWeakPtr> ConnectionPool::getConnectionAsync(
         const std::string& endpoint) {
     boost::unique_lock<boost::mutex> lock(mutex_);
-
+    PoolMap::iterator cnxIt = pool_.end();
     if (poolConnections_) {
-        PoolMap::iterator cnxIt = pool_.find(endpoint);
+        cnxIt = pool_.find(endpoint);
         if (cnxIt != pool_.end()) {
-            ClientConnectionPtr cnx = cnxIt->second.lock();
-
-            if (cnx && !cnx->isClosed()) {
-                // Found a valid or pending connection in the pool
-                LOG_DEBUG("Got connection from pool for " << endpoint << " use_count: "  //
-                        << (cnx.use_count() - 1) << " @ " << cnx.get());
-                return cnx->getConnectFuture();
-            } else {
-                // Deleting stale connection
-                LOG_INFO("Deleting stale connection from pool for " << endpoint << " use_count: "
-                        << (cnx.use_count() - 1) << " @ " << cnx.get());
-                pool_.erase(endpoint);
+            // endpoint exists in the map
+                ClientConnectionContainerPtr containerPtr = cnxIt->second;
+                if (containerPtr && containerPtr->isFull()) {
```

Comment:
@rdhabalia are you referring to the name of the mapped type?

whatever it's called, doesn't need to be a pointer since the only copy will happen when you insert the empty container in the map.